nailgun/entity_mixins.py

```diff
@@ -931,3 +937,325 @@ def update(self, fields=None):
+        * No other method should alter ``fields`` or ``query``.
+
+        :param results: A list of dicts, where each dict is a set of attributes
+            for one entity. The contents of these dicts are as is returned from
+            the server.
+        :returns: A list of dicts, where each dict is a set of attributes for
+            one entity. The contents of these dicts have been normalized and
+            can be used to instantiate entities.
+
+        """
+        fields = self.get_fields()
+        normalized = []
+        for result in results:
+            # For each field that we know about, copy the corresponding field
+            # from the server's search result. If any extra attributes are
+            # copied over, Entity.__init__ will raise a NoSuchFieldError.
+            # Examples of problematic results from server:
+            #
+            # * organization_id (denormalized OneToOne. see above)
+            # * organizations, organization_ids (denormalized OneToMany. above)
+            # * updated_at, created_at (these may be handled in the future)
+            # * sp_subnet (Host.sp_subnet is an undocumented field)
+            #
+            attrs = {}
+            for field_name, field in fields.items():
+                if isinstance(field, OneToOneField):
+                    try:
+                        attrs[field_name] = _get_entity_id(field_name, result)
+                    except MissingValueError:
+                        pass
+                elif isinstance(field, OneToManyField):
+                    try:
+                        attrs[field_name] = _get_entity_ids(field_name, result)
+                    except MissingValueError:
+                        pass
+                else:
+                    try:
+                        attrs[field_name] = result[field_name]
+                    except KeyError:
+                        pass
+            normalized.append(attrs)
+        return normalized
+
+    def search(self, fields=None, query=None, filters=None):
+        """Search for entities.
+
+        At its simplest, this method searches for all entities of a given kind.
+        For example, to ask for all :class:`nailgun.entities.Organization`
+        entities::
+
+            Organization().search()
+
+        Values on an entity are used to generate a search query, and the
+        ``fields`` argument can be used to specify which fields should be used
+        when generating a search query::
+
+            org = Organization(name='foo', description='bar')
+            results = org.search()  # Search by name and description.
+            results = org.search(fields={'name', 'description'})  # Same.
+            results = org.search(fields={'name'})  # Search by name.
+            results = org.search(fields={'description'})  # Search by desc
+            results = org.search(fields=set())  # Search for all organizations.
+            results = org.search(fields={'title'})  # Error!
```

Comment:
@omaciel nailed it.

> Here we have error, because in `Organization(name='foo', description='bar')` we did not specify `title` field value explicitly, right?

Correct. You're telling nailgun to create a search using the "title" attribute, but the "title" attribute is not set.

> and at that moment it has `None` value?

Incorrect. No `title` attribute has been set at all.

```python
>>> from nailgun.entities import Organization
>>> org = Organization(name='foo', description='bar').
>>> hasattr(org, 'title')
False
>>> org.title
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'Organization' object has no attribute 'title'
```

Further example:

```python
>>> from nailgun.entities import Organization
>>> org = Organization(name='foo', description='bar')                                                     
>>> org.search_payload()
{'description': 'bar', 'name': 'foo'}
>>> org.search_payload({'name'})
{'name': 'foo'}
>>> org.search_payload({'title'})                                                                         
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/ichimonji10/code/nailgun/nailgun/entity_mixins.py", line 1012, in search_payload
    value = getattr(self, field)
AttributeError: 'Organization' object has no attribute 'title'
```